gce/burst-training/census-analysis.py

```diff
@@ -0,0 +1,126 @@
+import argparse
+import joblib
+import numpy as np
+import os
+import pandas as pd
+import tensorflow as tf
+import scipy
+import sklearn
+import xgboost as xgb
+
+
+from sklearn.model_selection import RandomizedSearchCV
+
+
+def main(mode, census_data_path, model_output_path, cv_iterations=1):
+    TRAIN_DATA = '{}/adult.data'.format(census_data_path)
+    TEST_DATA = '{}/adult.test'.format(census_data_path)
+
+    COLUMNS = (
+        'age',
+        'workclass',
+        'fnlwgt',
+        'education',
+        'education-num',
+        'marital-status',
+        'occupation',
+        'relationship',
+        'race',
+        'sex',
+        'capital-gain',
+        'capital-loss',
+        'hours-per-week',
+        'native-country',
+        'income-level'
+    )
+
+    CATEGORICAL_COLUMNS = (
+        'workclass',
+        'education',
+        'marital-status',
+        'occupation',
+        'relationship',
+        'race',
+        'sex',
+        'native-country'
+    )
+
+    with tf.gfile.Open(TRAIN_DATA, 'r') as train_data:
+        train_raw_df = pd.read_csv(train_data, header=None, names=COLUMNS)
+    train_features_df = train_raw_df.drop('income-level', axis=1)
+    train_labels_df = (train_raw_df['income-level'] == ' >50K')
+
+    with tf.gfile.Open(TEST_DATA, 'r') as test_data:
+        test_raw_df = pd.read_csv(test_data, names=COLUMNS, skiprows=1)
+    test_features_df = test_raw_df.drop('income-level', axis=1)
+    test_labels_df = (test_raw_df['income-level'] == ' >50K.')
+
+    if mode == 'train':
+        encoders = {col:sklearn.preprocessing.LabelEncoder()
+                    for col in CATEGORICAL_COLUMNS}
+
+        for col in CATEGORICAL_COLUMNS:
+            train_features_df[col] = encoders[col].fit_transform(
+                train_features_df[col])
+
+        classifier_0 = xgb.XGBClassifier(nthread=-1)
+
+        classifier = RandomizedSearchCV(
+            classifier_0,
+            param_distributions={
+                'max_depth': [3,4,5,6,7],
+                'learning_rate': list(np.arange(0.01, 0.2, 0.01)),
+                'n_estimators': scipy.stats.poisson(100),
+                'gamma': scipy.stats.expon(scale=0.001),
+                'reg_lambda': scipy.stats.expon(scale=0.001)
+            },
+            n_iter=cv_iterations
+        )
+
+        classifier.fit(train_features_df, train_labels_df, verbose=10)
+
+        model_export = {
```

Comment:
suggestion: use sklearn's pipeline to chain encoders and classifier into a single estimator.  you still can use the whole pipeline in randomized search cv.

Reply:
Considered this, but it doesn't really work well with pandas dataframes - the column replacements do not fit well into the sklearn pipeline and to do it properly would require custom Estimator implementations which is not really the point of this work.

Reply:
What are column replacements?  

sklearn preprocessing has this general transformer, which might be used to wrap functions: http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html#sklearn.preprocessing.FunctionTransformer

Reply:
`FunctionTransformer` is difficult to use for stateful changes like this as making the encoder available in the transforming functions scope is non-trivial.

Reply:
The encoder state needs to be preserved across runs.

Reply:
I might be missing something here.  If the encoder is part of the pipeline, it would be fitted along with the rest of the pipeline, and its state would be joblib.dumped as well.

Reply:
Based on my understanding, the pattern to implement the preprocessor here with `FunctionTransformer` would be to create a new class with a `__call__` method which stored the column preprocessors in a member dictionary probably, where the class `fit` method called the `fit` methods of each of the individual column preprocessors.

I felt hesitant to do this because:

1. It felt like it would be too involved to create an estimator subclass acting as a compound preprocessor in that manner - it would definitely add a lot of requirements to this sample in terms of tests, etc.

2. It also felt like so many contortions around our model code would distract from the point of the example, which is the bash scripts automating the use of (probably preemptible) instances capable of computation on a large number of threads.

What do you think?